Added MarkupElementRewriter

ajaybhargavb opened the pull request:
#2584 

This is going to be run before and after the tag helper phase. I am currently not adding any errors at this level. All errors will be added during the tag helper phase. Sending this as a separate PR just to make sure everyone is on the same page.
Added tests

ajaybhargavb commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+{
+    internal class MarkupElementRewriter
+    {
+        public static RazorSyntaxTree Rewrite(RazorSyntaxTree syntaxTree)
```
This is called before tag helper phase.

rynowak replied:
Names are wacky here because there are two rewritiers. Suggest something like `AddElementNodes` `RemoveElementNodes`

ajaybhargavb commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+            return newSyntaxTree;
+        }
+
+        public static RazorSyntaxTree RemoveMarkupElement(RazorSyntaxTree syntaxTree)
```
This is called after tag helper phase.

ajaybhargavb commented on test/Microsoft.AspNetCore.Razor.Language.Test/Legacy/MarkupElementRewriterTest.cs:
```diff
@@ -0,0 +1,123 @@
+            BaselineTest(rewritten);
+
+            var unrewritten = MarkupElementRewriter.RemoveMarkupElement(rewritten);
+            Assert.Equal(syntaxTree.Root.SerializedValue, unrewritten.Root.SerializedValue);
```
Each test will also run the removal phase and make sure the output matches the original syntax tree. It didn't make sense to add separate tests for the removal phase.

rynowak replied:
Love it.

ajaybhargavb commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+                        continue;
+                    }
+
+                    var tagName = tagBlock.GetTagName();
```
`GetTagName()` is an extension method to `MarkupTagBlockSyntax`. This will change in the future when tag block have a first class TagName child.

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+
+namespace Microsoft.AspNetCore.Razor.Language.Syntax
+{
+    internal class MarkupElementRewriter
```
`static class`?

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+
+            private bool IsVoidElement(MarkupTagBlockSyntax tagBlock)
+            {
+                return VoidElements.Contains(tagBlock.GetTagName(), StringComparer.OrdinalIgnoreCase);
```
This is doing `Enumerable.Contains`. Did you mean to do `VoidElements.Contains(tagBlock.GetTagName())` since the hashset is already using an `OrdinalIgnoreCase` comparer?

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+                return VoidElements.Contains(tagBlock.GetTagName(), StringComparer.OrdinalIgnoreCase);
+            }
+
+            private bool IsSelfClosing(MarkupTagBlockSyntax tagBlock)
```
I'm guessing this is is a end tag like `</>` but don't see a corresponding test. Should this be unit tested?

ajaybhargavb replied:
There is a corresponding test. https://github.com/aspnet/Razor/pull/2624/files#diff-8dfd8aeca4f15b18bb45a38795fd1c02R94

pranavkm replied:
Wouldn't either of those have one or more child nodes?

ajaybhargavb replied:
They can't have any child nodes.

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+            {
+                var lastChild = tagBlock.ChildNodes().LastOrDefault();
+
+                return lastChild?.GetContent().EndsWith("/>", StringComparison.Ordinal) ?? false;
```
Out of curiosity, why does `IsEndTag` use literal tokens and this use `GetContent()`?

ajaybhargavb replied:
`IsEndTag` was a copy-paste-modify. I was wondering the same thing as why is it doing all the weird index stuff. But I left it as is for now.

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+                return false;
+            }
+
+            private SyntaxNode RewriteMalformedTags(SyntaxNode parent, int malformedTagCount)
```
This method is invoked when `malformedTagCount != _startTagTracker.Count`, but reading the code it looks like `_startTagTracker.Count must be > malformedTagCount` (since it's popped that many times). Does the `if` need to be changed?

ajaybhargavb replied:
Sure. It doesn't functionally change anything but I'll change the if.

ajaybhargavb replied:
Done

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+                    }
+                    else if (IsEndTag(tagBlock))
+                    {
+                        if (!string.IsNullOrEmpty(CurrentStartTagName) &&
```
You can remove this check. Next line should suffice since you already verify `tagName` has a value

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+                        node = RewriteNodeCore(node, startTag: tagBlock, tagChildren: Enumerable.Empty<RazorSyntaxNode>(), endTag: null);
+
+                        // Since we rewrote 'node', it's children are different. Update our collection.
+                        children = node.ChildNodes();
```
How does this work? Should this restart from the beginning since the children are different or continue from the next index because this node is already observed?

ajaybhargavb replied:
It won't restart. It just continues from where it left off. I still have to update the collection because rewriting node changes the children.

rynowak replied:
What if instead you created a new list of children and then rewrote it once.

So like:
```
A
  B 
  C 
  D
```

Iterate through the children of `A` - `(B, C, D)` and rewrite just those nodes as needed so you end up with `(B, C!, D)` (assume C was rewritten). Then rewrite `A` once to replace its children. That's how I would expect something like this to work.

ajaybhargavb replied:
The problem is that when I rewrite (B, C, D) => (B, C!, D), it actually becomes (B, C!, D!) because everything after a rewritten node is different because red nodes are created on demand

rynowak replied:
Can you explain this in more detail? I'm not sure that I get it.

ajaybhargavb replied:
Sorry I missed this part "created a **new** list of children". What I said doesn't apply in that case. I'll give this a shot

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+
+            public Rewriter()
+            {
+                _startTagTracker = new Stack<TagBlockTracker>();
```
Inline?

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+                MarkupTagBlockSyntax endTag)
+            {
+                var body = new SyntaxList<RazorSyntaxNode>(tagChildren);
+                var markupElement = SyntaxFactory.MarkupElement((MarkupTagBlockSyntax)startTag, body, endTag);
```
Can this be initialized inline (in the lambda)?

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+                }
+
+                // Replace nodes
+                var rewrittenElement = parent.ReplaceNodes(originalNodes, (original, rewritten) => {
```
Formatting. Also local functions for 💯 points

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,362 @@
+
+                var currentIndex = 0;
+                var children = node.ChildNodes();
+                while (currentIndex < children.Count)
```
`for (var i= 0; i < children.Count; i++)` works really well here.

ajaybhargavb commented:
Ping @NTaylorMullen @rynowak @pranavkm

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,352 @@
+            return newSyntaxTree;
+        }
+
+        private class Rewriter : SyntaxRewriter
```
Is there precendent for any of this? Like is this based on something or did you write it fresh?

rynowak replied:
https://github.com/aspnet/Blazor/blob/master/src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentDocumentRewritePass.cs FYI

ajaybhargavb replied:
> Is there precendent for any of this? Like is this based on something or did you write it fresh?

This is mostly fresh. Some of the malformed tag handling logic is very similar to  https://github.com/aspnet/Razor/blob/master/src/Microsoft.AspNetCore.Razor.Language/Legacy/TagHelperParseTreeRewriter.cs#L730

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,352 @@
+                if (node != null)
+                {
+                    node = RewriteNode(node);
+                }
```
Can a node ever be null?

ajaybhargavb replied:
Yes. `Visit()` is called on every child and `Visit()` calls `node.Accept(this)` if the node is not null.

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,352 @@
+                    node = RewriteNode(node);
+                }
+
+                return base.Visit(node);
```
Does it make sense for this to be a preorder traversal instead? https://github.com/aspnet/Blazor/blob/master/src/Microsoft.AspNetCore.Blazor.Razor.Extensions/ComponentDocumentRewritePass.cs#L55

ajaybhargavb replied:
I'll take a look

ajaybhargavb replied:
That works as well

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,352 @@
+                {
+                    // Nothing to replace
+                    return parent;
+                }
```
Does this preserve order correctly? What about a case like:
```
<div>Hi</div>
@Something!
<div>bye</div>
```

rynowak replied:
I see, you did the below.

rynowak replied:
I think this would make a little more sense if you mutated `originalNodes` in place and preserved ordering while doing so.

rynowak commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,352 @@
+                    return true;
+                }
+
+                // Could not recover tag. Aka we found an end tag without a corresponding start tag.
```
Make sure you have a test for recovery when the cause is a malformed void element. We had a bug for this recently in Blazor.
```
<input ....></input>
```

ajaybhargavb replied:
Added.

ajaybhargavb commented:
🆙 📅

pranavkm commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,329 @@
+                    TrackChild(rewritten, rewrittenChildren);
+                }
+
+                tagChildren.ForEach(c => TrackChild(null, rewrittenChildren));
```
It'd also let you pass in `Array.Empty<RazorSyntaxNode>()` instead of `new List<>`

NTaylorMullen commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,329 @@
+                    return node;
+                }
+
+                _startTagTracker.Clear();
```
In practice is this ever not already clear?

ajaybhargavb replied:
It should be. Just being extra defensive.

NTaylorMullen commented on src/Microsoft.AspNetCore.Razor.Language/Syntax/MarkupElementRewriter.cs:
```diff
@@ -0,0 +1,329 @@
+                }
+            }
+
+            private void TrackChild(SyntaxNode child, List<SyntaxNode> rewrittenChildren)
```
Confusing af

ajaybhargavb replied:
Added this to https://github.com/aspnet/Razor/issues/2619.

ajaybhargavb commented:
I've addressed most of the feedback here. I've added the one I didn't address to this list https://github.com/aspnet/Razor/issues/2619 to be addressed later. Merging this now.
